arch/arm/core/aarch64/reset.S

```diff
@@ -63,6 +66,56 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
 	msr	spsel, #0
 
 	ldr	x0, =(z_interrupt_stacks)
+
+#define IS_POWER_OF_TWO(X) ((X) && (((X) & ((X) - 1)) == 0))
+#if !IS_POWER_OF_TWO(ARCH_STACK_PTR_ALIGN)
+	#error "This code assumes ARCH_STACK_PTR_ALIGN is a power of 2"
+	/*
+	 * How to implement this if necessary:
+	 * 1. Introduce a macro that expands to
+	 *    a. `and ... val-1` if val is a power of two
+	 *    b. `mod ... val` if val is not a power of two
+	 * (Why a and b instead of just b? Because mod is _very_ slow
+	 *  compared to and.)
+	 * 2. Use the macro for operations on ARCH_STACK_PTR_ALIGN
+	 */
+#endif
+#if (CONFIG_ISR_STACK_SIZE % ARCH_STACK_PTR_ALIGN)
+	#error "This code assumes that the stack size is aligned"
+	/*
+	 * How to implement this if necessary:
+	 * 1. Round down the resulting stack size to the next aligned value.
+	 * 2. Do all size checks / etc on the resulting rounded-down value.
+	 * (However, if your stack size isn't a multiple of `ARCH_STACK_PTR_ALIGN`
+	 * you're just wasting memory needlessly.)
+	 */
+#endif
+
+	/*
+	 * This check should be compile-time, but better late than never.
+	 * This check also could be `b.ne .` instead of `b.eq` over a `b .`,
+	 * but this way the resulting program counter value when spinning
+	 * is unambiguously in the assert.
+	 * This check could call something other than just spin, but this
+	 * is so early in boot that we would be very unlikely to recover.
+	 */
+
+	/* assert(z_interrupt_stacks % ARCH_STACK_PTR_ALIGN == 0) */
+#ifdef CONFIG_ASSERT
+	ands x1, x0, #(ARCH_STACK_PTR_ALIGN - 1)
+	b.eq z_interrupt_stacks_align_ok
+	/* If you're spinning here check the alignment of z_interrupt_stacks.
+	 * You could work around this by rounding `z_interrupt_stacks` up
+	 * to the next multiple of ARCH_STACK_PTR_ALIGN, but that would
+	 * just be wasting space.
+	 */
+	b .
```

Comment:
I am not sure I like the spinning here - it gives little information to the developer. I understand, though, that we are too early in the boot sequence and proper cpu fault reporting mechanism is not established.

Also, it is not clear why this is under CONFIG_ASSERT; after all, there is no explicit assert statement here.

I'd like others to comment here (@carlocaione @andyross ) too; IMO, we should be using build time checks for ISR stack size/alignment (and any other thread stack definition).

Unless I am missing something, defining thread and isr stack sizes using the right kernel macros (and their arch-specific extensions) should do the work, so no need for run-time checks.

Reply:
> I am not sure I like the spinning here - it gives little information to the developer.

I am in agreement here. It's mainly a spin here because I don't know of a better alternative:

1. C (and ASM) does not consider the address of a global a compile-time-known construct, so I cannot check this at compile time without resorting to non-standard constructs like `__alignof__`.
2. Checking this later in runtime does not accomplish anything, as by the time we would check we'd _already_ have jumped off into the weeds (likely a Stack Alignment fault)
3. At this point in bootflow an actual assert call will do nothing other than obscure the source of the failure.

(Also: is the expectation for Zephyr that someone will be debugging early boot without access to a debug probe of some sort? With a debug probe this gives fairly clear information to the developer so far as I can tell. We're sitting at a branch-to-self, find where in the source the address corresponds to, see the comment in the source immediately before it as to what this is checking and why...)

> it is not clear why this is under CONFIG_ASSERT; after all, there is no explicit assert statement here.

It's conceptually an assert, and would be an actual assert if I had any confidence at this point in bootflow that an assert would do anything other than further obscure the actual failure. Unfortunately...

Is there a better pattern I should use for this sort of early-boot check?

> Unless I am missing something, defining thread and isr stack sizes using the right kernel macros (and their arch-specific extensions) should do the work, so no need for run-time checks.

Fair enough. I tend to be fairly overzealous about runtime checks in asserts-enabled builds in places that aren't performance-sensitive. (i.e. ideally _both_ a compile-time and a runtime check, with the compile-time check always enabled and the runtime check enabled for `CONFIG_ASSERT`.).

This is mainly because I've had too many cases in the past where a compile-time check got inadvertently broken or removed (e.g. moved to a file that isn't compiled in a specific build) and I then spent a bunch of time trying to rewind time and figuring out what happened before things jumped off into the weeds.